cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,78 +282,370 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+	// we don't have ubuntu-data host to get the unsealed "bare" key, so
+	// we have to unlock with the sealed one from ubuntu-seed
+	saveFallbackKey := filepath.Join(boot.InitramfsEncryptionKeyDir, "ubuntu-save.recovery.sealed-key")
+	saveDevice, isEncryptedDev, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(ctx.disk, "ubuntu-save", saveFallbackKey, true)
 	if err != nil {
+		if isEncryptedDev {
+			logger.Noticef("failed to find or unlock encrypted ubuntu-save partition: %v", err)
+			ctx.degradedState.SaveState = "enc-not-found"
+		} else {
+			// either catastrophic error or inconsistent disk, if
+			// ubuntu-data was an encrypted device, then ubuntu-save
+			// must be also
+			logger.Noticef("cannot unlock ubuntu-save: %v", err)
+			ctx.degradedState.SaveState = "other"
+		}
+
+		return nil, errStateDone
+	}
+
+	ctx.degradedState.SaveKey = "fallback"
+	ctx.saveDevice = saveDevice
+
+	return m.mountSave, nil
+}
+
+func (m *stateMachine) mountSave(ctx *context) (stateFunc, error) {
+	// TODO: should we fsck ubuntu-save ?
+	if err := doSystemdMount(ctx.saveDevice, boot.InitramfsUbuntuSaveDir, nil); err != nil {
+		logger.Noticef("error mounting ubuntu-save from partition %s: %v", ctx.saveDevice, err)
+		ctx.degradedState.SaveState = "not-mounted"
+	} else {
+		// if we couldn't verify whether the mounted save is valid, bail out of
+		// the state machine and exit snap-bootstrap
+		if err := verifyMountPointCtx(ctx, boot.InitramfsUbuntuSaveDir, "ubuntu-save"); err != nil {
+			return nil, err
+		}
+
+		ctx.degradedState.SaveState = "mounted"
+	}
+
+	return nil, errStateDone
+}
+
+func generateMountsModeRecover(mst *initramfsMountsState) error {
+	// steps 1 and 2 are shared with install mode
+	if err := generateMountsCommonInstallRecover(mst); err != nil {
 		return err
 	}
-	if !matches {
-		return fmt.Errorf("cannot validate boot: ubuntu-data mountpoint is expected to be from disk %s but is not", disk.Dev())
+
+	// get the disk that we mounted the ubuntu-seed partition from as a
+	// reference point for future mounts
+	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+	if err != nil {
+		return err
 	}
-	if haveSave {
-		// 3.2a we have ubuntu-save, verify it as well
-		matches, err = disk.MountPointIsFromDisk(boot.InitramfsUbuntuSaveDir, diskOpts)
+
+	// 3. run the state machine logic for mounting partitions, this involves
+	//    trying to unlock then mount ubuntu-data, and then unlocking and
+	//    mounting ubuntu-save
+	//    see the state* functions for details of what each step does and
+	//    possible transition points
+
+	// TODO: should we just put a diagram here explaining the states too?
+
+	ctx := &context{
+		disk:          disk,
+		degradedState: recoverDegradedState{},
+	}
+
+	// first state to execute is to unlock ubuntu-data with the run key
+	machine := newStateMachine()
+	for {
+		err := machine.execute(ctx)
 		if err != nil {
+			if err == errStateDone {
```

Comment:
That is slightly strange, this is how the machine tells it has reached the end state? I think I would slightly prefer something like: 
```
finished, err := machine.execute(mctx)
```